fossilize_db.cpp

```diff
@@ -739,6 +792,9 @@ struct StreamArchive : DatabaseInterface
 		if (!file)
 			return false;
 
+		/* we are seeking into the file, only reading tiny chunks: disable readahead */
+		LINUX_fadvise(file, 0, 0, POSIX_FADV_RANDOM);
```

Comment:
New version of commit now disables readahead instead of discarding cache on each read.

Reply:
After letting it run over night, it seems to be more memory/cache friendly as swap usage only increased by 200 MB instead of 2 GB.

Reply:
This seems like it would be brutal when reading a database from a cold HDD. Is there any issue with scanning through metadata with SEQUENTIAL, then transition the file into RANDOM or NOREUSE?

Reply:
Transitioning should be fine... NOREUSE is a no-op, see my other [comment](https://github.com/ValveSoftware/Fossilize/pull/101#discussion_r510175874).

Reply:
Well, I mean, if we start out from RANDOM, like this PR is doing, then the initial scan through the archive will be very slow. We observed this on Windows in particular, where O_SEQUENTIAL from a cold HDD **really** helped.

What I'm thinking is, the ideal way of doing this would be:
- SEQUENTIAL, while scanning through. Ideally, any pages we have read through will be de-prioritized so that scanning sequentially over a huge archive does not dominate the cache.
- RANDOM after the initial scan. If pages are dropped while we scan through the archive, we should not do any read-ahead while processing the archive.

Reply:
Possibly, throw in a DONTNEED for the whole file between SEQUENTIAL and RANDOM if that matters.

Reply:
Okay, so if the tests showed that, I'll conclude that the seeks are sub-page or sub-readahead-window size, and SEQUENTIAL really helps here. So your plan sounds good.

While scanning through the entries sequentially, we could just DONTNEED what's behind of the seek pointer (so we do not run into an issue with sub-page sizes), right? However, if the process starts working with the DB later, it may need to re-read blocks from disk. Apparently, DONTNEED wouldn't de-prioritize the cached pages, it will throw them away almost immediately (unless some CPU or thread still has a hold on it).

Reply:
Hm, right. If DONTNEED immediately drops the pages, that's bad. We *really* want to keep the pages in cache if it fits. We just don't want to dominate other processes if it doesn't fit. I guess the best we can do is SEQUENTIAL -> RANDOM and leave it at that.

When I figure out a shared SHM metadata block, this initial scan is not necessary in child processes, and we should be able to just go RANDOM out of the gate.

Reply:
> When I figure out a shared SHM metadata block

Maybe we should wait for that change then first before fiddling around in this PR any further? It sounds like it has the best potential of reducing memory pressure. In the end, we may be fine with using just WILLNEED in the parent before child spawning, and DONTNEED after the children are done, and use RANDOM for the actual file access. WILLNEED will trigger immediate readahead of the complete file using device optimal IO size.

Reply:
I guess we can park this PR until then, yes.

Reply:
I do not claim credit for the fadvise lines, so feel free to incorporate that into your code redesign when it makes sense. I think the work here has given us enough clues and a better understanding how the kernel works with fadvise.